### PR TITLE
feat: program registry

### LINKED ISSUE #148: `Program` deployment and associated `register_program` contract registry predicate.
Once the new program `Predicate` type is ready to land with the updated `Program` DAG model, the contract registry should be updated in the node to allow `Solution`s to register `Program`s independently of `Predicate`s.

`Program`s can be stored using a unique key within the existing contract registry.

`Program`s will need to be deployed either in the same `Solution` that their dependent `Predicate`(s) are deployed in, or in any earlier `Solution`.

The contract registry mutation for deploying a program will likely look something like this:

```yaml
key: [2, <program-ca>]
value: [len_bytes, <program-bytecode-padded>]
```

where `2` is the `programs` storage prefix within the contract registry.

## Validation

Program deployment validation will involve ensuring:

1. The content address that appears in the key matches the actual program content address (i.e. `sha256` of the program bytecode).
2. The length in bytes does not exceed `Program::MAX_SIZE`.

### PR DESCRIPTION
**Pull request by supiket**
## Program registry
- [x] Program registry contract and program deployment
- [x] Tests

Like the contract registry contract, the program registry contract only has storage and no logic in the predicate for now. Related and next steps: #48, https://github.com/essential-contributions/essential-base/issues/190 and https://github.com/essential-contributions/essential-base/issues/221.

#### Review tip
Main change is to `BigBang` in `node-types/src/lib.rs`. Registry contract's layout is described here. In turn, function `validate_inner` in `crates/node/src/validate.rs` validates the programs.

## Update to latest base 
That means adapting to three major changes: DAG model of programs, https://github.com/essential-contributions/essential-base/pull/229 and https://github.com/essential-contributions/essential-base/pull/236.

#### Review tip 
The changes to `node-db-sql` and `node-db` are related solely to this. 

Closes #148

### CODE REVIEW

File: crates/node-types/src/lib.rs
Comment (supiket) [inline crates/node-types/src/lib.rs:259]: - [ ] Actually it is not but it should be

**Review by mitchmindtree — COMMENTED** [review]
AWESOME work! Totally nailed it :mage_woman: 

It looks like we might need to `cargo update -p essential-check` to update the `essential-base` crates in the Cargo.lock file and pull in the latest changes including `Solution` naming etc.

Otherwise lgtm!

File: crates/node-types/src/lib.rs
Comment (supiket) [reply crates/node-types/src/lib.rs:69]: Oh, I missed that the `0` and `1` prefixes in contract registry were due to the index of the corresponding map in the storage block. Thanks! 

Added 0 prefix for programs.

File: crates/node/src/test_utils.rs
Comment (supiket) [reply crates/node/src/test_utils.rs:174]: Agreed

File: crates/node-types/src/lib.rs
Comment (supiket) [reply crates/node-types/src/lib.rs]: Added a comment that links to the relevant issue

**Comment by supiket** [comment]
One rename left after latest commit: 

- [ ] Rename all mentions of dec_vars in db to predicate_data

**Comment by nfurfaro** [comment]
> One rename left after latest commit:
> 
> * [ ]  Rename all mentions of dec_vars in db to predicate_data

perfect, this will unblock me nicely.

File: crates/node-db-sql/src/lib.rs
Comment (mitchmindtree) [inline crates/node-db-sql/src/lib.rs:114]: if we want to keep these in alphabetical order it looks like we might wanna move this const too

**Review by mitchmindtree — APPROVED** [review]
Wow. HUGE work on the renaming - great job on also going though all the SQL and comments too!

This lgtm. It sounds like it's time to land these `essential-base` `develop` changes.

I don't think the `builder` update will be too major (certainly much smaller than this), so in retrospect its probably fine to open a PR for it after we land this.

I'll make the final tweaks to `essential-base/develop` now and get it ready for review and publish before merging into `main`, then we can change this to point back to crates.io :+1:

**Review by nfurfaro — APPROVED** [review]
Looks good! That's a lot of renaming and SQL.

File: crates/node-db-sql/sql/insert/block_solution_set.sql
Comment (supiket) [reply crates/node-db-sql/sql/insert/block_solution_set.sql:6]: Opened #167

File: Cargo.toml
Comment (mitchmindtree) [inline Cargo.toml:23]:
```suggestion
essential-check = "0.10.0"
essential-hash = "0.8.0"
essential-sign = "0.8.0"
essential-types = "0.6.0"
```

**Review by mitchmindtree — APPROVED** [review]

**Pull request merged**
